rq/worker.py

```diff
@@ -1499,6 +1499,8 @@ def handle_job_success(self, job: 'Job', queue: 'Queue', started_job_registry: S
 
                     pipeline.execute()
 
+                    assert job.started_at
+                    assert job.ended_at
```

Comment:
I don't personally like adding assertions just to satisfy type checkers because it's causing us to run extra code during run time. Can we just use type ignore?

Reply:
If there's some version of `type ignore` that lets us bypass just the `None` type warning, it would be even better.

Reply:
Assertions aren't exactly expensive so of any runtime code I think they're probably fine, but I understand your perspective. `# type: ignore[union-attr]` is a bit of a sledgehammer and may silence multiple errors where it's possible one of them is actually an error, conversely using `# type: ignore` does not update mypy's perspective of the code and you need to use `# type: ignore` on every access rather than having mypy narrow its understanding of what possible values the variable can be.

The "other version" is to use `cast` which is also runtime code (also a function call) and in order for mypy to narrow its understanding of what the variable is you'd need to reassign the value. This works for variables, but not for attributes like this (or at least I don't think it does). Regardless a reassign + cast is more expensive than an assertion.

I would argue the reason to use assertions would be to make sure that type definitions really do match the runtime data model. If you don't run anything at runtime then you just are relying on luck, and that you didn't lie to the typechecker. Adding assertions allows tests (runtime) to check if you're actually typing something too strictly and what you believed was an assertion is incorrect.